### PR TITLE
Add conflict parameter annotation and a map for accessing the conflict objects

### PR DESCRIPTION
**Pull request by no2chem**
This patch adds the @ConflictParameter annotation, which indicates to the runtime
that a parameter should be used to generate objects for fine-grained conflict
resolution. 

This patch does not implement fine-grained conflict resolution, which will have to
be implemented in both the transactional context and the sequencer.

### CODE REVIEW

File: runtime/src/main/java/org/corfudb/runtime/collections/ISMRMap.java [hunk 1]
```diff
@@ -133,8 +133,28 @@
...
+     * @param value value to be associated with the specified key
+     * @throws UnsupportedOperationException if the <tt>put</tt> operation
+     *                                       is not supported by this map
+     * @throws ClassCastException            if the class of the specified key or value
+     *                                       prevents it from being stored in this map
+     * @throws NullPointerException          if the specified key or value is null
+     *                                       and this map does not permit null keys or values
+     * @throws IllegalArgumentException      if some property of the specified key
+     *                                       or value prevents it from being stored in this map
+     */
```
Comment (dahliamalkhi) [inline runtime/src/main/java/org/corfudb/runtime/collections/ISMRMap.java]: First, I believe that key is a (write-) conflict parameter for a blindPut, no?

Second, how do we handle conflict resolution between two transaction on the same object, one of which has specific conflict parameters, the other has no explicit ones?

File: annotations/src/main/java/org/corfudb/runtime/object/ICorfuSMR.java [hunk 2]
```diff
@@ -29,7 +29,9 @@
     /** Get a map from strings (function names) to undoRecord methods.
      * @return The undo record map. */
     Map<String, IUndoRecordFunction<T>> getCorfuUndoRecordMap();
-
+    /** Get a map from strings (function names) to conflict methods.
+     * @return The undo record map. */
```
Comment (dahliamalkhi) [inline annotations/src/main/java/org/corfudb/runtime/object/ICorfuSMR.java]: return comment is a leftover?

File: annotations/src/main/java/org/corfudb/runtime/object/IConflictFunction.java [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+
+/** A function which is used to calculate the fine-grained conflict set.
+ * Created by mwei on 12/15/16.
+ */
+public interface IConflictFunction {
+    /** Given the parameters to a function, calculate the conflict set.
+     * @param args  The arguments to the function.
+     * @return      The fine-grained conflict set.
+     */
+    Object[] calculate(Object... args);
```
Comment (dahliamalkhi) [inline annotations/src/main/java/org/corfudb/runtime/object/IConflictFunction.java]: not a meaningful name, please change.

File: runtime/src/main/java/org/corfudb/runtime/collections/ISMRMap.java (hunk 1 above)
Comment (no2chem) [inline runtime/src/main/java/org/corfudb/runtime/collections/ISMRMap.java]: 1) blindPut doesn't have an upcall, so it doesn't have a conflictParameter.
2) No conflictParameter means always conflict (conflicts against everything).

**Pull request merged**